### PR TITLE
Return a user friendly error on an empty query

### PR DESCRIPTION
**Pull request by AramZS**
This assures that if the user hits the namespace with no query they are
presented with a friendly well-formatted error. Noted in ticket #200

### CODE REVIEW

**Comment by AramZS** [comment]
@jasonbahl Finished writing unit tests, this is ready for code review.

**Comment by AramZS** [comment]
hmm, after some checking I don't think that the errors here are a result of my changes?

**Comment by jasonbahl** [comment]
@AramZS ya, looking into it. Something with the router tests is acting up.

File: tests/test-json-responses.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
+<?php
+/**
+ * Test case for Basic Router responses.
+ *
+ * @group ajax
+ */
+class WPGraphQL_Test_Router extends WP_Ajax_UnitTestCase {
```
Comment (jasonbahl) [inline tests/test-json-responses.php]: @AramZS looks like the issue is that you used the same class name here as we have here: https://github.com/wp-graphql/wp-graphql/blob/develop/tests/test-router.php#L3

Try changing this class name to something like `WP_GraphQL_Test_JSON_Responses`

File: tests/test-json-responses.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+		/**
+		 * Set the query var to "graphql" so we can mock like we're visiting the endpoint via
+		 */
+		set_query_var( 'graphql', true );
+		$GLOBALS['wp']->query_vars['graphql'] = true;
+
+		/**
+		 * Instantiate the router
+		 */
+		$router = new \WPGraphQL\Router();
```
Comment (jasonbahl) [inline tests/test-json-responses.php]: Looks like this is never used, as the use of it is commented out?

File: tests/test-json-responses.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+		add_action( 'wp_ajax_graphql_resolve_http_request', [ '\WPGraphQL\Router', 'resolve_http_request' ] );
+		/**
+		 * Process the request using our filtered data
+		 */
+		try {
+			$this->_handleAjax( 'graphql_resolve_http_request' );
+		} catch ( WPAjaxDieContinueException $e ) {}
+		// $router::resolve_http_request();
```
Comment (jasonbahl) [inline tests/test-json-responses.php]: Can remove this if it's not needed

**Comment by jasonbahl** [comment]
@AramZS it also looks like the tests in `test-router.php` that test the `resolve_http_request` cause the tests to stop running altogether. 

test-router-abrupt-end

**Comment by AramZS** [comment]
Weird. Ok, I'll check in to it!

File: tests/test-json-responses.php (hunk 2 above)
Comment (AramZS) [reply tests/test-json-responses.php]: This gets used by the hook and I think it needs to be instantiated earlier.

**Comment by AramZS** [comment]
@jasonbahl I fixed up the tests and figured out a way to unobtrusively make the router capable of responding on AJAX but not other non-header ready requests. 

The end point of that though was that I realized that the router function doesn't actually complete and push out a response on any of the tests in the router file. I saw other things are being checked, but I wasn't sure if that was the intent. That said, I fixed some changes I made previously to respect the current method. Those were the issues that were breaking the router test.

**Comment by AramZS** [comment]
I still am getting `xdebug_get_headers()` related errors, but those are the ones that you said to ignore right?

**Comment by jasonbahl** [comment]
@AramZS ya, we can ignore those for now. . .as those are environment specific (I just need to make sure the Travis instances have xdebug). . .not sure why those tests sometimes pass and sometimes fail though

**Comment by jasonbahl** [comment]
> The end point of that though was that I realized that the router function doesn't actually complete and push out a response on any of the tests in the router file. I saw other things are being checked, but I wasn't sure if that was the intent. That said, I fixed some changes I made previously to respect the current method. Those were the issues that were breaking the router test.

Right now the intent was to test individual functions rather than full integration tests. . .where _most_ of our testing is more "integration" testing, where we're not testing individual functions on their own, but testing queries which map to several functions. 

Probably would be good to separate "integration" and "unit" tests at some point...

**Comment by jasonbahl** [comment]
@AramZS I'm still getting funky output from the test, but it's passing now. . .not sure it's an issue, just haven't seen that before 🤔 

see: https://travis-ci.org/wp-graphql/wp-graphql/jobs/268428149

**Comment by AramZS** [comment]
@jasonbahl I know where it is coming from, now when DOING_AJAX is true and `headers_sent` is true it executes the `wp_die` inside `wp_send_json` whereas before if the `headers_sent` was true it would never send a response (that's what I was talking about above). I believe that the headers are marked as set in the router test class, which is why it never outputted into the buffer. I noticed that from the CLI I never got this error in the tests I ran, but I see it is getting it when I run the full suite. 

Is there anywhere else besides the tests I created (which have output buffering built in, which is why they don't show that output) where DOING_AJAX is defined as true or the headers are set as part of the test? That must be where the echoed buffer is coming from.

**Comment by jasonbahl** [comment]
> Is there anywhere else besides the tests I created (which have output buffering built in, which is why they don't show that output) where DOING_AJAX is defined as true or the headers are set as part of the test? That must be where the echoed buffer is coming from.

No. . .but I *think* we can tag the test with `@runInSeparateProcess` and that _might_ solve it?

**Comment by AramZS** [comment]
@jasonbahl The thing is I don't think it is the `WPGraphQL_JSON_Responses` test. That one run on its own doesn't echo the buffer, so it is some other test causing the rogue output.

**Comment by AramZS** [comment]
Yeah, I see it, the issue is that the DOING_AJAX const effects some other place where it is calling the response down the line. Adding the `runInSeparateProcess` directive.

**Comment by jasonbahl** [comment]
@AramZS thanks for working on this! looks great.

**Pull request merged**